Source/WebCore/rendering/RenderThemeCocoa.mm

```diff
@@ -78,11 +169,133 @@ @implementation WebCoreRenderThemeBundle
+
+void RenderThemeCocoa::drawAxialGradient(CGContextRef context, IOSGradientRef gradient, const FloatPoint& startPoint, const FloatPoint& stopPoint, Interpolation interpolation)
+{
+    RetainPtr<CGShadingRef> shading = adoptCF(CGShadingCreateAxial(sRGBColorSpaceRef(), startPoint, stopPoint, getSharedFunctionRef(gradient, interpolation), false, false));
+    CGContextDrawShading(context, shading.get());
+}
+
+void RenderThemeCocoa::drawRadialGradient(CGContextRef context, IOSGradientRef gradient, const FloatPoint& startPoint, float startRadius, const FloatPoint& stopPoint, float stopRadius, Interpolation interpolation)
+{
+    RetainPtr<CGShadingRef> shading = adoptCF(CGShadingCreateRadial(sRGBColorSpaceRef(), startPoint, startRadius, stopPoint, stopRadius, getSharedFunctionRef(gradient, interpolation), false, false));
+    CGContextDrawShading(context, shading.get());
+}
+
 bool RenderThemeCocoa::shouldHaveCapsLockIndicator(const HTMLInputElement& element) const
 {
     return element.isPasswordField();
 }
 
+Color RenderThemeCocoa::pictureFrameColor(const RenderObject& buttonRenderer)
+{
+    return systemColor(CSSValueAppleSystemControlBackground, buttonRenderer.styleColorOptions());
+}
+
+Color RenderThemeCocoa::backgroundImageColor(const RenderObject& buttonRenderer)
+{
+    return systemColor(CSSValueWebkitControlBackground, buttonRenderer.styleColorOptions());
+}
+
+void RenderThemeCocoa::paintFileUploadIconDecorations(const RenderObject&, const RenderObject& buttonRenderer, const PaintInfo& paintInfo, const IntRect& rect, Icon* icon, FileUploadDecorations fileUploadDecorations)
+{
+    GraphicsContextStateSaver stateSaver(paintInfo.context());
+
+    IntSize cornerSize(kThumbnailBorderCornerRadius, kThumbnailBorderCornerRadius);
+
+    Color pictureFrameColor = this->pictureFrameColor(buttonRenderer);
+
+    IntRect thumbnailPictureFrameRect = rect;
+    IntRect thumbnailRect = rect;
+    thumbnailRect.contract(2 * kThumbnailBorderStrokeWidth, 2 * kThumbnailBorderStrokeWidth);
+    thumbnailRect.move(kThumbnailBorderStrokeWidth, kThumbnailBorderStrokeWidth);
+
+    if (fileUploadDecorations == MultipleFiles) {
+        // Smaller thumbnails for multiple selection appearance.
+        thumbnailPictureFrameRect.contract(kMultipleThumbnailShrinkSize, kMultipleThumbnailShrinkSize);
+        thumbnailRect.contract(kMultipleThumbnailShrinkSize, kMultipleThumbnailShrinkSize);
+
+        // Background picture frame and simple background icon with a gradient matching the button.
+        Color backgroundImageColor = this->backgroundImageColor(buttonRenderer);
+        paintInfo.context().fillRoundedRect(FloatRoundedRect(thumbnailPictureFrameRect, cornerSize, cornerSize, cornerSize, cornerSize), pictureFrameColor);
+        paintInfo.context().fillRect(thumbnailRect, backgroundImageColor);
+        {
+            GraphicsContextStateSaver stateSaver2(paintInfo.context());
+            CGContextRef cgContext = paintInfo.context().platformContext();
```

Comment:
I'm confused. How is this OK with DOMGPUP on?

Reply:
what do you mean?

Reply:
I mean that "calling platformContext() and using CG in the Web process is usually an insta-crash with DOMGPUP on, so how is this working?"

Reply:
is DOMGPUP enabled by default? Because otherwise I have no idea how it's working 😅

Reply:
It is on iOS, yes.

Reply:
ya I have no idea then, since it works fine on both iOS and macOS 😅 I'll try testing again

Reply:
I just tested it again, and it works. And Also none of this code is new for iOS